src/data/sum/order.lean

```diff
@@ -0,0 +1,209 @@
+-/
+
+namespace sum
+variables {α β γ δ : Type*}
+
+/-! ### Unbundled relation classes -/
+
+section lift_rel
+variables (r : α → α → Prop) (s : β → β → Prop)
+
+instance [is_refl α r] [is_refl β s] : is_refl (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro (a | a), exacts [lift_rel.inl (refl _), lift_rel.inr (refl _)] }⟩
+
+instance [is_irrefl α r] [is_irrefl β s] : is_irrefl (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ (⟨a, _, h⟩ | ⟨a, _, h⟩); exact irrefl _ h }⟩
+
+instance [is_trans α r] [is_trans β s] : is_trans (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, c, hbc⟩ | ⟨_, c, hbc⟩),
+  exacts [lift_rel.inl (trans hab hbc), lift_rel.inr (trans hab hbc)] }⟩
+
+instance [is_antisymm α r] [is_antisymm β s] : is_antisymm (α ⊕ β) (lift_rel r s) :=
+⟨by { rintro _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, _, hba⟩ | ⟨_, _, hba⟩); rw antisymm hab hba }⟩
+
+end lift_rel
+
+section lex
+variables (r : α → α → Prop) (s : β → β → Prop)
+
+instance [is_refl α r] [is_refl β s] : is_refl (α ⊕ β) (lex r s) :=
+⟨by { rintro (a | a), exacts [lex.inl (refl _), lex.inr (refl _)] }⟩
+
+instance [is_irrefl α r] [is_irrefl β s] : is_irrefl (α ⊕ β) (lex r s) :=
+⟨by { rintro _ (⟨a, _, h⟩ | ⟨a, _, h⟩); exact irrefl _ h }⟩
+
+instance [is_trans α r] [is_trans β s] : is_trans (α ⊕ β) (lex r s) :=
+⟨by { rintro _ _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, c, hbc⟩ | ⟨_, c, hbc⟩),
+  exacts [lex.inl (trans hab hbc), lex.sep _ _, lex.inr (trans hab hbc), lex.sep _ _] }⟩
+
+instance [is_antisymm α r] [is_antisymm β s] : is_antisymm (α ⊕ β) (lex r s) :=
+⟨by { rintro _ _ (⟨a, b, hab⟩ | ⟨a, b, hab⟩) (⟨_, _, hba⟩ | ⟨_, _, hba⟩); rw antisymm hab hba }⟩
+
+instance [is_total α r] [is_total β s] : is_total (α ⊕ β) (lex r s) :=
+⟨begin
+  rintro (a | a) (b | b),
+  { exact (total_of r a b).imp lex.inl lex.inl },
+  { exact or.inl (lex.sep _ _) },
+  { exact or.inr (lex.sep _ _) },
+  { exact (total_of s a b).imp lex.inr lex.inr }
+end⟩
+
+instance [is_trichotomous α r] [is_trichotomous β s] : is_trichotomous (α ⊕ β) (lex r s) :=
+⟨begin
+  rintro (a | a) (b | b),
+  { exact (trichotomous_of r a b).imp3 lex.inl (congr_arg _) lex.inl },
+  { exact or.inl (lex.sep _ _) },
+  { exact or.inr (or.inr $ lex.sep _ _) },
+  { exact (trichotomous_of s a b).imp3 lex.inr (congr_arg _) lex.inr }
+end⟩
+
+instance [is_well_order α r] [is_well_order β s] : is_well_order (α ⊕ β) (sum.lex r s) :=
+{ wf := sum.lex_wf is_well_order.wf is_well_order.wf }
+
+end lex
+
+/-! ### Disjoint sum of two orders -/
+
+section disjoint
+
+instance [has_le α] [has_le β] : has_le (α ⊕ β) := ⟨lift_rel (≤) (≤)⟩
+
+instance [preorder α] [preorder β] : preorder (α ⊕ β) :=
+{ le_refl := λ _, refl _,
+  le_trans := λ _ _ _, trans,
+  .. sum.has_le }
+
+instance [partial_order α] [partial_order β] : partial_order (α ⊕ β) :=
+{ le_antisymm := λ _ _, antisymm,
+  .. sum.preorder }
+
+end disjoint
+
+/-! ### Linear sum of two orders -/
+
+namespace lex
+
+localized "attribute [-instance] sum.has_le" in lex
```

Comment:
Do we have the lemma that says the disjoint order is ≤ the lexicographic order?

Reply:
No. I can't (easily) state it currently because I still don't have the `lex` synonym. This is what the TODO is about.

Reply:
Oh wait, I forgot the TODO.

Reply:
Now we do!